resource_aws_kinesis_firehose_delivery_stream: Avoid invalid state on refresh

staffan-einarsson opened the pull request:
When refreshing the state of a aws_kinesis_firehose_delivery_stream
resource, the AWS SDK may have more that one way to describe the same
configuration. In the case of ProcessingConfiguration, this can lead to
perpetual diffs and for DataFormatConversionConfiguration it can lead to
invalid state that causes later refresh and plan operations to crash.

We have solved these two situations here by detecting the alternative
representation of the "disabled" configuration, and disambiguated the
terraform state data.

<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

Fixes #8982 and possibly also #6053 and #4392

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/aws_kinesis_firehose_delivery_stream: After manual edit in AWS console, refresh no longer causes invalid state file.
```

staffan-einarsson commented:
There are some failing acceptance tests, but these seem to fail on `master` as well so I don't think they are related to this.

```
$ make testacc TESTARGS='-run=TestAccAWSKinesisFirehoseDeliveryStream'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSKinesisFirehoseDeliveryStream -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]

...

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidProcessorType (43.06s)
    testing.go:561: Step 0, expected error:

        errors during apply: error creating Kinesis Firehose Delivery Stream: ValidationException: 1 validation error detected: Value 'NotLambda' at 'extendedS3DestinationConfiguration.processingConfiguration.processors.1.member.type' failed to satisfy constraint: Member must satisfy enum value set: [Lambda]
        	status code: 400, request id: d481983e-5529-3a6a-8100-f9830facd9f8

        To match:

        must be 'Lambda'

...

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidParameterName (54.27s)
    testing.go:561: Step 0, expected error:

        errors during apply: error creating Kinesis Firehose Delivery Stream: ValidationException: 1 validation error detected: Value 'NotLambdaArn' at 'extendedS3DestinationConfiguration.processingConfiguration.processors.1.member.parameters.1.member.parameterName' failed to satisfy constraint: Member must satisfy enum value set: [NumberOfRetries, RoleArn, BufferIntervalInSeconds, LambdaArn, BufferSizeInMBs]
        	status code: 400, request id: c12609b6-7538-d2ce-94a7-687f5369cd1a

        To match:

        must be one of 'LambdaArn', 'NumberOfRetries'

...

FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1023.624s
make: *** [testacc] Error 1
```

bflad commented:
The fixes for the testing are in https://github.com/terraform-providers/terraform-provider-aws/pull/8886

mattk42 commented:
I believe as part of this fix we also need to mark the input_format_configuration, output_format_configuration, and schema_configuration sections of the parsing schema as optional (as in my case they are not sent along with the "Enabled: False" DataFormatConversionConfiguration state. 

I have submitted a PR into @staffan-einarsson's branch to see if we can make that part of this PR.

bflad commented:
Is it possible to use a Terraform configuration to get the API response into the state the console generates? To ensure we're properly fixing this and prevent future regressions for this behavior, we'll likely want to add to our acceptance testing.

staffan-einarsson commented:
@bflad, I think it comes down to choosing one form of the API request/response to be canonical from terraform's point of view. In the firehose API, if we call CreateDeliveryStream with either of the following payloads, we get the same thing.
```
{
  "DeliveryStreamName": "brandslangen",
  "DeliveryStreamType": "DirectPut",
  "ExtendedS3DestinationConfiguration": {
    "BucketARN": "arn:aws:s3:::hinken.supercorp.io",
    "BufferingHints": {
      "IntervalInSeconds": 300,
      "SizeInMBs": 5
    },
    "CompressionFormat": "UNCOMPRESSED",
    "EncryptionConfiguration": {
      "NoEncryptionConfig": "NoEncryption"
    },
    "Prefix": "",
    "RoleARN": "arn:aws:iam::REDACTED:role/rollen",
    "S3BackupMode": "Disabled"
  }
}
```

```
{
  "DeliveryStreamName": "brandslangen",
  "DeliveryStreamType": "DirectPut",
  "ExtendedS3DestinationConfiguration": {
    "BucketARN": "arn:aws:s3:::hinken.supercorp.io",
    "BufferingHints": {
      "IntervalInSeconds": 300,
      "SizeInMBs": 5
    },
    "CompressionFormat": "UNCOMPRESSED",
    "DataFormatConversionConfiguration": {
      "Enabled": false
    },
    "EncryptionConfiguration": {
      "NoEncryptionConfig": "NoEncryption"
    },
    "Prefix": "",
    "ProcessingConfiguration": {
      "Enabled": false
    },
    "RoleARN": "arn:aws:iam::REDACTED:role/rollen",
    "S3BackupMode": "Disabled"
  }
}
```
However, in contrast to the usual behaviour around default values, DescribeDeliveryStream remembers which form was used in the Create call and returns that form. So we get different descriptions depending on the form used in the last Create or Update call.

Looking at this now it is clear that the API gravitates towards the forms `"DataFormatConversionConfiguration":{"Enabled":false}` and `"ProcessingConfiguration":{"Enabled":false}`, because once these forms have been put in once, we can't get them out even if the UpdateDestination call does not include them. I'm leaning towards terraform always using these forms as a way to minimise noise.

However, in the terraform state I still think it makes sense that we represent the absence of these configurations with empty lists, not the API form with `enabled: false` and that the sub-attributes are Required (as they currently are for `data_format_conversion_configuration` but not for `processing_configuration`.) The benefit of having them Required is to avoid configurations that make little sense but the drawback is that now, if you are affected by this bug, there is no super smooth way to get the state repaired.

@mattk42, you are correct that by making the `data_format_conversion_configuration` sub-attributes optional, state will self heal, but it comes at the cost of less precise configuration validation going forward. Using the fix from this PR the state can be mended by reimporting the aws_kinesis_firehose_delivery_stream:
```
terraform state rm aws_kinesis_firehose_delivery_stream.brandslangen
terraform import aws_kinesis_firehose_delivery_stream.brandslangen arn:aws:firehose:eu-west-1:REDACTED:deliverystream/brandslangen
```
This works becuase the problem is in parsing the bad state, not in refreshing it from the API.

@bflad, I wanted to add some acceptance testing but I was a bit at a loss on how to approach it, since the core issue is that the resource is mutated outside of terraform. I'd be happy to work on it if you have any hints on that.

mattk42 commented:
@staffan-einarsson Agreed with your thoughts on my changes and closed my PR into your branch. I was misunderstanding the purpose of the schema. Thanks for the help!

staffan-einarsson commented:
Hmm, I tried to rebase this onto current `master` but it seems the PR was closed instead. Can't seem to reopen it either. I've updated the logic for Create and Update so that we always put the firehose in a state that is close to what we would get after a console modification. I also added acceptance testing for the external modification.

Can this be opened again or should I create a new PR?

bflad commented:
It appears the branch this was using was deleted:

> @staffan-einarsson staffan-einarsson deleted the staffan-einarsson:b-aws_kinesis_firehose_delivery_stream-inconsistent_state  branch 4 hours ago

Which should explain why it cannot be reopened (even by a maintainer). You'll need to submit a new pull request. 👍

staffan-einarsson commented:
Yeah, sorry about that. 😅
I'll open up a new one.
